src/basic/headers-util.c

```diff
@@ -0,0 +1,235 @@
+
+/* We follow bash for the character set. Different shells have different rules. */
+#define VALID_HEADER_NAME_CHARS \
+        DIGITS LETTERS "_"      \
+                       "-"
+
+static bool headers_name_is_valid_n(const char *e, size_t n) {
+
+        if (n == SIZE_MAX)
+                n = strlen_ptr(e);
+
+        if (n <= 0)
+                return false;
+
+        assert(e);
+
+        /* POSIX says the overall size of the environment block cannot be > ARG_MAX, an individual assignment
+         * hence cannot be either. Discounting the equal sign and trailing NUL this hence leaves ARG_MAX-2 as
+         * longest possible variable name. */
+        if (n > (size_t) sysconf(_SC_ARG_MAX) - 2)
+                return false;
+
+        for (const char *p = e; p < e + n; p++)
+                if (!strchr(VALID_HEADER_NAME_CHARS, *p))
+                        return false;
+
+        return true;
+}
+
+bool headers_name_is_valid(const char *e) {
+        return headers_name_is_valid_n(e, strlen_ptr(e));
+}
+
+bool headers_value_is_valid(const char *e) {
+        if (!e)
+                return false;
+
+        if (!utf8_is_valid(e))
+                return false;
+
+        return true;
+}
+
+bool headers_assignment_is_valid(const char *e) {
+        const char *eq;
+
+        eq = strchr(e, '=');
+        if (!eq)
+                return false;
+
+        if (!headers_name_is_valid_n(e, eq - e))
+                return false;
+
+        if (!headers_value_is_valid(eq + 1))
+                return false;
+
+        return true;
+}
+
+bool strv_headers_is_valid(char **e) {
+        STRV_FOREACH(p, e) {
+                size_t k;
+
+                if (!headers_assignment_is_valid(*p))
+                        return false;
+
+                /* Check if there are duplicate assignments */
+                k = strcspn(*p, "=");
+                STRV_FOREACH(q, p + 1)
+                        if (strneq(*p, *q, k) && (*q)[k] == '=')
+                                return false;
+        }
+
+        return true;
+}
+
+bool strv_headers_name_is_valid(char **l) {
+        STRV_FOREACH(p, l) {
+                if (!headers_name_is_valid(*p))
+                        return false;
+
+                if (strv_contains(p + 1, *p))
+                        return false;
+        }
+
+        return true;
+}
+
+bool strv_headers_name_or_assignment_is_valid(char **l) {
+        STRV_FOREACH(p, l) {
+                if (!headers_assignment_is_valid(*p) && !headers_name_is_valid(*p))
+                        return false;
+
+                if (strv_contains(p + 1, *p))
+                        return false;
+        }
+
+        return true;
+}
+
+char *convert_header(char *src) {
+        assert(src);
+        int len = strlen_ptr(src) + 2;
+        char *dest = malloc(len);
+        char *s = src;
+        char *d = dest;
+        do {
+                if (*s == '=') {
+                        *d++ = ':';
+                        *d++ = ' ';
```

Comment:
this is going to be  memory corruption if the input string contains multiple `=`